timemachine/ff/amber_converter.py

```diff
@@ -0,0 +1,313 @@
+                    elif cn.nodeName == "Bond" and args.method == "harmonic_bond":
+                        # generate BCCs based on harmonic_bond atom types, note this "over-symmetrizes"
+                        # i.e. symmetry classes are too large, and some just look grossly wrong, eg. serine
+                        src_idx = int(cn.attributes["from"].value)
+                        dst_idx = int(cn.attributes["to"].value)
+
+                        # get the atom types
+                        src_type = atom_idxs_to_types[src_idx]
+                        dst_type = atom_idxs_to_types[dst_idx]
+                        bond_src_type = atom_types_name_to_class[src_type]
+                        bond_dst_type = atom_types_name_to_class[dst_type]
+                        # canonicalize bond_type
+                        bond_src_type, bond_dst_type, src_idx, dst_idx = dual_sort(
+                            bond_src_type, bond_dst_type, src_idx, dst_idx
+                        )
+                        key_tuple = (bond_src_type, bond_dst_type)
+                        assert key_tuple in hb_bond_set
+
+                        if key_tuple not in bond_type_map:
+                            bond_type_map[key_tuple] = len(bond_type_map)
+
+                        bond_param_idxs.append(bond_type_map[key_tuple])
+
+                        bond_list.append((src_idx, dst_idx))
+                        bond_iota += 1
+
+                    elif cn.nodeName == "Bond" and args.method == "template_bond":
+                        # looks only at the template bond types, but this "under-symmetrizes"
+                        src_idx = int(cn.attributes["from"].value)
+                        dst_idx = int(cn.attributes["to"].value)
+
+                        # get the atom types
+                        src_type = atom_idxs_to_types[src_idx]
+                        dst_type = atom_idxs_to_types[dst_idx]
+
+                        # canonicalize bond_type
+                        bond_src_type, bond_dst_type, src_idx, dst_idx = dual_sort(src_type, dst_type, src_idx, dst_idx)
+                        bond_key = (bond_src_type, bond_dst_type)
+
+                        if bond_key not in bond_type_map:
+                            bond_type_map[bond_key] = len(bond_type_map)
+
+                        bond_param_idxs.append(bond_type_map[bond_key])
+
+                        bond_list.append((src_idx, dst_idx))
+                        bond_iota += 1
+
+                    elif cn.nodeName == "Bond" and args.method == "nonbonded":
+                        # generate BCCs based on nonbonded atom types formed by the concatenation of q,sig,eps
+                        # this roughly generates the correct symmetries (probably at the cost of having more parameters)
+                        src_idx = int(cn.attributes["from"].value)
+                        dst_idx = int(cn.attributes["to"].value)
+                        src_type = atom_idxs_to_types[src_idx]
```

Comment:
nit: The xml is always ordered Atoms -> Bonds ? Since this reads from the dict written to in the if Atom branch.

Reply:
Yes - I'll add a comment for this.